### ISSUE TITLE
Inconsistency in saving biochemistry object

### ISSUE THREAD

**Issue by samseaver** [ModelSEED/ModelSEED#157]
I'm a little concerned about "saving" biochemistry objects under the same reference.  I feel like I am getting inconsistent results, and it has reduced me to the point at which to be confident in my biochemistry objects, I use the --saveas option every time, and every time I re-run my scripts for merging the biochemistries, I delete the store and start over.

We've had this conversation before, but I wanted to write and ask if any testing has been done to show that modification/addition of a large number of objects into a biochemistry object which is then saved _to the same reference_ has been done?

**Comment by devoid** [comment]
Not sure what you mean when you say:

> if any testing has been done to show that modification/addition of a large number of objects into a biochemistry object which is then saved to the same reference has been done?

Whenever you save a biochemistry object the following happens:
1. MS Object code calls serializeToDB.
2. The database potentially changes the UUID of the provenance object being saved.
3. The database updates the existing alias to point to the new UUID, or creates the alias if it didn't exist already.
4. The database updates the MS object with the new UUID

At no point is the "old object" altered in the database. If you record the UUIDs for the biochemistry as you go along, you can always recover them. There's also `ms history` which will list the entire history of an alias or UUID reference.

**Comment by cshenry** [comment]
Yes, and it absolutely positively works. However, you need to understand how it works.

For example, right now you may get weird results if you do this:

1.) Build biochemistry
2.) Build a model linked to biochemistry
3.) Change biochemistry
4.) Try to use the changed biochemistry in your model

This will not work, because the link to the biochemistry in the model still points to the old biochemistry.
We are working on changing that link from a UUID to a reference to overcome this problem… but this isn't done yet.

**Comment by samseaver** [comment]
Thanks Scott, this is an additional layer of knowledge that is new to me, and it helps reduce the uncertainty in my mind.

**Comment by samseaver** [comment]
I'm re-opening this issue, because there is definitely an inconsistency.  Last time I experienced this, I attributed it to switching branches or running while changing code, etc.  But I can honestly say that the following commands (`ms bio mergebio` run with the same options each time) leads to a different biochemistry each time:

```
bash-3.2$ ms bio mergebio Cns_KEGG Cns_MetaCyc -n KEGG,MetaCyc -m Search_InChI,StereoFreeSearch_InChI,searchname --verbose -a Merged > Merged.out 2>&1
bash-3.2$ ms list biochemistry/seaver/Merged
compounds            (16955)
aliasSets            (10)
reactions            (8674)
compartments         (13)
bash-3.2$ ms bio mergebio Cns_KEGG Cns_MetaCyc -n KEGG,MetaCyc -m Search_InChI,StereoFreeSearch_InChI,searchname --verbose -a Merged > Merged.out 2>&1
bash-3.2$ ms list biochemistry/seaver/Merged
compounds            (25291)
aliasSets            (16)
reactions            (18112)
compartments         (13)
ancestor_uuids       (1)
bash-3.2$ ms bio mergebio Cns_KEGG Cns_MetaCyc -n KEGG,MetaCyc -m Search_InChI,StereoFreeSearch_InChI,searchname --verbose -a Merged > Merged.out 2>&1
bash-3.2$ ms list biochemistry/seaver/Merged
compounds            (16955)
aliasSets            (10)
reactions            (8674)
compartments         (13)
```

I can confirm that I did not do any messing around "behind scenes" here, just run the same command over, and then `ms list` to see the result.  As you can see, the second time I run it, there is a significantly different number of compounds and reactions in the object.  The second time I run it, is when I get the desired results, the other two times I run it, the first and third time, I only get KEGG objects.

What is happening is that I am merging KEGG and MetaCyc biochemistries, and for some reason, the MetaCyc compounds and reactions are only saved into the final biochemistry object (Merged) if I run it twice.

@cshenry I suspect strongly, that this code in MS::Biochemistry::mergeBiochemistry() is the culprit:

```
$bio->updateLinks($type,$uuidTranslation,1,1);
$bio->_clearIndex();
$self->updateLinks($type,$uuidTranslation,1,1);
$self->_clearIndex();
```

The same code works when using `ms bio consolidatebio`, but in this case, a new biochemistry object is created before merging, so there's effectively no uuids and links in the new object.  However, in `ms bio mergebio`, this isn't the case, and so I think MS::Biochemistry::updateLinks() still isn't doing the job properly.

**Comment by cshenry** [comment]
Very curious behavior. I will take a look...

**Comment by samseaver** [comment]
I've been able to narrow down the behavior, by trying to update the UUID of a single object using this code:

```
my $obj=$new_biochem->getObject($new_set->attribute(),$new_uuid);

 print "Current UUID before updating: ",$obj->uuid(),"\n";

 $obj->uuid($old_uuid);

 print "Current UUID after first update: ",$obj->uuid(),"\n";

 $store->save_object("biochemistry/".$auth->username()."/".$Biochemistry."_Test",$new_biochem);
```

The output shows that the object's uuid is actually updated, but when I use `ms list`, I can only retrieve the new_uuid, and not the old_uuid, so this means that the problem somehow lies in the MS::Store::save_object() function.

**Comment by samseaver** [comment]
I just did another test on this behaviour:

```
    my $obj=$new_biochem->getObject($new_set->attribute(),$new_uuid);

    print "Current UUID before updating: ",$obj->uuid(),"\n";

    $obj->uuid($old_uuid);

    print "Current UUID after first update: ",$obj->uuid(),"\n";

    $obj=$new_biochem->getObject($new_set->attribute(),$new_uuid);

    print "Object after retrieval using new_uuid: ",$obj,"\n";

    $obj=$new_biochem->getObject($new_set->attribute(),$old_uuid);

    print "Object after retrieval using old_uuid: ",$obj,"\n";
```

The output fails to print the Object hash the second time, so this extra code shows that I'm unable to retrieve the object from the biochemistry using its supposedly updated UUID, only with its old UUID, so in a sense I see that the culprit is the uuid() function, in that updating the uuid doesn't translate back into the biochemistry, unless I'm misunderstanding how it works.

**Comment by cshenry** [comment]
Sam,

This makes sense. There's nothing magical here. The biochemistry maintains an index of the uuids, and that index is not updated when you change the uuid on an object. This would be very hard, and undesirable to fix.

Essentially, you shouldn't be manually extracting objects and altering their uuids and expecting everything to work. UUIDs should only be altered in the context of a function that understands the full consequences of making that alteration and deals with it, as the update links function does. Remember, everything links to everything else in the Model SEED via the UUIDs. They should not be altered without extreme care.

So my question is, why are you altering the uuid on sub objects? What is it that you're trying to do?

Chris

**Comment by samseaver** [comment]
I'm literally trying to update the object UUIDs of my biochemistry object with the UUIDs from the current ModelSEED biochemistry, like you asked me to.  I'm using the MS::Biochemistry::mergeBiochemistry() function as my guide.

You're using `$object->uuid($new_uuid)` within that function, and that function is giving us the weird behavior, as previously discussed, so what I'm saying is that I'm trying to update uuids within a biochemistry object, and I'm geting similar behavior, and I'm trying to nail down why I get that behavior (i.e. I have to run the same code twice to get the uuids to update properly).
